README.md

## Desafio Alelo

Criar uma API (tema livre) com:

    - Get/Post/Delete/Put;
    - Testes unitários, usando banco em memória;
    - Teste “mockando” o retorno de uma API de terceiros;
    - Documentação via Swagger;

Caso você não domine algum item do desafio proposto, não desista: envie tudo o que você sabe. O objetivo é avaliar o seu jogo de cintura, seu nível de conhecimento e a sua criatividade. A API deve ser simples, objetiva e aplicar boas práticas de desenvolvimento.

### Implantação
O projeto está configurado para build totalmente automatizada via Maven e pode ser executado via fat-jar ou deploy em servidor por meio de WAR.

Build via jar
>mvn package

Build para deploy em servidor de aplicação
>mvn compile war:war

Deploy via Docker
> mvn package

> docker build -t silvioricardo/desafio-alelo:1.0-SNAPSHOT .

> docker run -d -p 8080:8080 silvioricardo/desafio-alelo:1.0-SNAPSHOT

### Documentação Técnica
Localizada em ['doc/documentacao.pdf'](doc/documentacao.pdf).

### Swagger

http://localhost:8080/swagger-ui.html

### Postman Collection

https://www.getpostman.com/collections/467e821404278e92a550

### Modelagem
Localizada em ['doc/modelagem.png'](doc/modelagem.png) ou neste arquivo no item Modelo.


### Base de dados
O banco de dados utilizado no projeto é H2, a conexão por meio de pool e configurado para persistência em arquivo local em 'data/Banco.db', utilizando JPA.

O script para geração de tabelas manualmente está no arquivo ['doc/schema.sql'](doc/schema.sql).

**Modelo**
![](doc/modelagem.png)


### O Web Service
O Web Service utiliza o padrão REST de retorno para o processamento das requisições enviadas:

    200 - Sucesso
    201 - Registro criado, incluído com sucesso.
    204 - Sem conteúdo para retornar
    400 - Dados enviados em formato incorreto
    404 - Registro não encontrado para a operação
    409 - Conflito, já existe registro.
    500 - Erro genérico no servidor, verificar aplicação

### Implantação
O projeto está configurado para build totalmente automatizada via Maven

A base de dados é H2 e com persistência local na pasta '.data/Banco.db' já inicializada.

Build via jar
>mvn package

Build para deploy em servidor de aplicação
>mvn compile war:war

### Payload Json para Testes
O Web Service utiliza o padrão REST de retorno para o processamento das requisições enviadas:

    200 - Sucesso
    201 - Registro criado, incluído com sucesso.
    204 - Sem conteúdo para retornar
    400 - Dados enviados em formato incorreto
    404 - Registro não encontrado para a operação
    409 - Conflito, já existe registro.
    500 - Erro genérico no servidor, verificar aplicação

**Usuário** /usuarios [GET/POST/PUT/DELETE]
>[GET] Retorna lista de usuários

>[POST] Adiciona um usuário
```json
{
  "email": "email@email.com",
  "endereco": "ENDEREÇO DO CARINHA",
  "nome": "Carinha da Silva",
  "senha": "hahahhaha",
  "telefone": "9899999233"
}
```

>[PUT] /{id} Atualiza um usuário
```json
{
  "email": "email@email.com",
  "endereco": "Novo endereço",
  "senha": "novasenha"
}
```

>[DELETE] /{id} Apaga um usuário com base na id /{id}

**Tarefas** /tarefas [GET/POST/PUT/DELETE]
>[GET] Retorna lista de tarefas

>[POST] Adiciona uma lista de tarefas
```json
{
  "descricao": "Tomar água",
  "usuarioId": 1
}
```

>[PUT] /{id} Atualiza uma tarefa com base na id
```json
{
  "dataConcluida": "2021-01-20T20:15:45.093Z",
  "dataPrevista": "2021-01-20T20:16:45.093Z",
  "id": 1,
  "status": 1,
  "listaId": 1,
  "usuarioId": 1
}
```

>[DELETE] /{id} Apaga uma tarefa com base na id

**Listas** /listas [GET/POST/PUT/DELETE]
>[GET] Retorna as listas

>[POST] Adiciona uma nova lista
```json
{
  "descricao": "Compras",
  "usuarioId": 1
}
```

>[PUT] /{id} Atualiza uma lista com base na id
```json
{
  "descricao": "Sonhos",
  "usuarioId": 1
}
```

>[DELETE] /{id} Apaga uma tarefa com base na id

CHANGELOG.MD
# 0.1.0
Versão inicial da aplicação


src/main/java/br/com/silvioricardo/alelo/todo/controller/UsuarioController.java
package br.com.silvioricardo.alelo.todo.controller;

import br.com.silvioricardo.alelo.todo.model.Usuario;
import br.com.silvioricardo.alelo.todo.repository.UsuarioRepository;
import br.com.silvioricardo.alelo.todo.service.UsuarioService;
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiResponse;
import io.swagger.annotations.ApiResponses;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

import static org.springframework.http.ResponseEntity.*;

/**
 * Controller com o CRUD para os endpoints do Usuario
 *
 */
@RestController
@AllArgsConstructor
@Api(value = "CRUD para Usuarios")
@RequestMapping({"/usuarios"})
@CrossOrigin(origins = "*")
public class UsuarioController {
    private final UsuarioService usuarioService;
    private final UsuarioRepository usuarioRepository;

    /**
     * Endpoint que retorna a lista de usuários
     *
     * STATUS:
     *  200 Sucesso
     *  204 Sem conteúdo
     *  500 Erro ao processar a solicitação
     *
     * @return JSON usuario
     */
    @ApiResponses(value = {
        @ApiResponse(code = 200, message = "Retorna a lista de usuários"),
        @ApiResponse(code = 204, message = "Nenhum usuário cadastrado"),
        @ApiResponse(code = 500, message = "Erro ao processar a requisição"),
    })
    @ApiOperation(value = "Retorna os usuários cadastrados", produces="application/json")
    @GetMapping
    public ResponseEntity<List<Usuario>> findAll(){
        List<Usuario> listaUsuarios = usuarioRepository.findAll();

        if(listaUsuarios.isEmpty())
            return noContent().build();

        return ok(listaUsuarios);
    }

    /**
     * Endpoint que retorna o usuário com base na id
     *
     * STATUS:
     *  200 Sucesso
     *  404 Não encontrado
     *  500 Erro ao processar a solicitação
     *
     * @return JSON usuario
     */
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = "Retorna o usuário informado com base na Id"),
            @ApiResponse(code = 404, message = "Não foi encontrado usuário com a Id informada"),
            @ApiResponse(code = 500, message = "Erro ao processar a requisição"),
    })
    @GetMapping("/{id}")
    @ApiOperation(value = "Procura usuário com base na Id", produces="application/json")
    public ResponseEntity<Usuario> findById(@PathVariable Long id){
        return ok(usuarioService.findById(id));
    }

    /**
     * Endpoint para inserção de novo usuario
     *
     * STATUS:
     *  (201) Criado
     *  (400) Requisição mal formatada
     *  (409) Conflito, entidade já existe na base de dados
     *  (500) Erro interno no servidor
     *
     * @param request String com o json da requisição
     * @return Response STATUS
     */
    @ApiResponses(value = {
            @ApiResponse(code = 201, message = "Usuário cadastrado com sucesso"),
            @ApiResponse(code = 400, message = "Requisição mal formatada"),
            @ApiResponse(code = 409, message = "Requisição mal formatada"),
            @ApiResponse(code = 500, message = "Erro ao processar a requisição"),
    })
    @PostMapping
    @ApiOperation(value = "Cadastra um novo usuário")
    public ResponseEntity<Usuario>cadastrar(@Valid @RequestBody Usuario usuario){
        Usuario novoUsuario = usuarioService.save(usuario);
        return new ResponseEntity<>(novoUsuario, HttpStatus.CREATED);
    }

    /**
     * Endpoint para atualização do usuário com base na Id recebida
     *
     * STATUS:
     *  (200) Sucesso
     *  (204) Sem conteúdo para resposta
     *  (404) Registro não encontrado ou inválido
     *
     * @param RequestBody Usuario Json da requisição
     * @param Long id do Usuário a ser atualizado
     * @return Response STATUS
     */
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = "Usuário atualizado com sucesso"),
            @ApiResponse(code = 400, message = "Requisição mal formatada"),
            @ApiResponse(code = 500, message = "Erro ao processar a requisição"),
    })
    @PutMapping("/{id}")
    @ApiOperation(value = "Atualiza dados do usuário")
    public ResponseEntity<Usuario> atualizar(@Valid @RequestBody Usuario usuario, @PathVariable Long id){
        if(usuario == null) {
            return new ResponseEntity<>(usuario, HttpStatus.BAD_REQUEST);
        }

        usuario.setId(id);
        Usuario usuarioAtualizado = usuarioService.update(usuario);
        return new ResponseEntity<>(usuarioAtualizado, HttpStatus.CREATED);
    }

    /**
     * Endpoint para remoção de um registro
     *
     * STATUS:
     *  (200) Sucesso
     *  (404) Registro não encontrado ou inválido
     *  (500) Erro ao processar a solicitação
     *
     * @param id Id do usuario
     * @return Response STATUS
     */
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = "Usuário apagado com sucesso"),
            @ApiResponse(code = 404, message = "Não foi encontrado usuário para a Id informada"),
            @ApiResponse(code = 500, message = "Erro ao processar a requisição"),
    })
    @DeleteMapping("/{id}")
    @ApiOperation(value = "Deleta um usuário")
    public ResponseEntity<?>deletar(@PathVariable Long id){
        return usuarioService.delete(id);
    }
}
